tests/conftest.py

```diff
@@ -62,18 +75,405 @@ def conn_params(test_session):
     yield test_session.get_bind().url
 
 
+#######################################################################
+# Test fixtures for code dependent on modelmeta database
+
+# TODO: Factor out common engine creation
+
+@pytest.fixture(scope='session')
+def mm_database_dsn():
+    """Test-session-wide testing.Postgresql instance; returns dsn for it"""
+    with testing.postgresql.Postgresql() as pg:
+        yield pg.url()
+
+
+@pytest.fixture(scope='session')
+def mm_engine(mm_database_dsn):
+    """Test-session-wide database engine"""
+    engine = create_engine(mm_database_dsn)
+    engine.execute("create extension postgis")
+    engine.execute(CreateSchema('test_meta'))
+    modelmeta.Base.metadata.create_all(bind=engine)
+    yield engine
+    engine.dispose()
+
+
+@pytest.fixture(scope='function')
+def mm_empty_session(mm_engine):
+    """Single-test database session.
+    All session actions are rolled back on teardown"""
+    session = sessionmaker(bind=mm_engine)()
+    session.execute('SET search_path TO test_meta, public')
+    yield session
+    session.rollback()
+    session.close()
+
+
+# Database test objects
+
+# Overview
+#
+# We set up the following objects which can be added to the database:
+#
+# model_1: Model
+#
+# emission_1: Emission
+# emission_2: Emission
+#
+# run_11: Run(model_1, emission_1)
+# run_12: Run(model_1, emission_2)
+#
+# data_file_1: DataFile(run_11)
+# data_file_2: DataFile(run_12)
+# data_file_3: DataFile(run_11)
+#
+# variable_alias_1: VariableAlias
+# variable_alias_2: VariableAlias
+#
+# dfv_dsg_time_series_11:
+#   DataFileVariableDSGTimeSeries(data_file_1, variable_alias_1)
+# dfv_dsg_time_series_12:
+#   DataFileVariableDSGTimeSeries(data_file_1, variable_alias_2)
+# dfv_dsg_time_series_21:
+#   DataFileVariableDSGTimeSeries(data_file_2, variable_alias_2)
+# dfv_dsg_time_series_31:
+#   DataFileVariableDSGTimeSeries(data_file_3, variable_alias_2)
+#
+# ensemble_1: Ensemble
+#   dfv_dsg_time_series_11
+#   dfv_dsg_time_series_21
+#
+# ensemble_2: Ensemble
+#   dfv_dsg_time_series_21
+#   dfv_dsg_time_series_31
+
+
+# Model
+
```

Comment:
The problem that explicit id's solve is:
- Objects have null id in code until they are flushed to the database (and/or refreshed).
- We wish to establish relationships between objects that are [transient (brand new and not yet flushed) or detached (removed from a session)](https://docs.sqlalchemy.org/en/13/orm/session_basics.html#adding-new-or-existing-items).
- We need id's to establish (some) relationships between objects: [SQLAlchemy relationships](https://docs.sqlalchemy.org/en/13/orm/relationships.html) that would (I think) let us avoid using explicit id's do not always exist in the ORM definition.

So, to simplify matters, I plumped for specifying id's everywhere. Suboptimal but not horror-ridden, and the explicit id parameter gives an easy way of providing unique values for attributes like `name` etc.

Reply:
Finally, I have a sneaking suspicion that since a fixture is a factory, if we don't specify id's explicitly per object, then each time we refer to the object we may get a new object with a new id, which is definitely not desired. And a similar consideration for any attribute that must be unique, either in the schema or for sane usage.

Reply:
I take this back. The necessary relationships *did* exist (they were just hard to find in a couple of cases). And we *do* have to create new objects each time, for reasons to be explained in the comments. So we can remove explicit id's, which simplified the code.